Envoy TLS support with header imposition

jrajahalme opened the pull request:
Add API support for TLS proxying for HTTP, with support for header modifications and (k8s) secrets.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9486)
<!-- Reviewable:end -->

aanm commented on test/vagrant-local-start.sh:
```diff
@@ -0,0 +1,31 @@
+
+set -e
+
+export K8S_VERSION=${K8S_VERSION:-1.16}
```
if #9661 is merged first, this needs to be changed to `1.17`

jrajahalme replied:
Sure, will do when #9661 has merged :-)

jrajahalme replied:
Issued #9681 to fix this.

aanm commented on pkg/defaults/defaults.go:
```diff
@@ -326,4 +326,8 @@ const (
+
+	// CertsDirectory is the default directory used to find certificates
+	// specified in the L7 policies.
+	CertsDirectory = "/var/run/cilium/certs"
```
`CertsDirectory = RuntimePath + "/certs"`

jrajahalme replied:
Changed.

aanm commented on pkg/k8s/annotate.go:
```diff
@@ -118,3 +121,20 @@ func (k8sCli K8sClient) AnnotateNode(nodeName string, encryptKey uint8, v4CIDR,
+		return nil, err
+	}
+	return result.Data, nil
+}
```
This also requires the following change:
```patch
diff --git a/install/kubernetes/cilium/charts/agent/templates/clusterrole.yaml b/install/kubernetes/cilium/charts/agent/templates/clusterrole.yaml
index d3b4c169f..a49ac1297 100644
--- a/install/kubernetes/cilium/charts/agent/templates/clusterrole.yaml
+++ b/install/kubernetes/cilium/charts/agent/templates/clusterrole.yaml
@@ -49,6 +49,12 @@ rules:
   - list
   - watch
   - update
+{{- if eq "k8s" .Values.global.tls.secretsBackend }}
+- apiGroups:
+  - ""
+  resources:
+  - secrets
+{{- end }}
 - apiGroups:
   - cilium.io
   resources:
diff --git a/install/kubernetes/cilium/values.yaml b/install/kubernetes/cilium/values.yaml
index 500ea8249..e17ac0f79 100644
--- a/install/kubernetes/cilium/values.yaml
+++ b/install/kubernetes/cilium/values.yaml
@@ -301,3 +301,6 @@ global:
 
   daemon:
     runPath: "/var/run/cilium"
+
+  tls:
+    secretsBackend: local
```

for which the users will be able to enable the RBAC for secrets access using

```
helm template cilium \
         --namespace=kube-system \
         --set global.tls.secretsBackend=k8s \
         ....
```

jrajahalme replied:
This would negate the need for `CertificatesPath`, assuming we would map a secret `<namespace>/<name>` to a path in `CertsDirectory` when `secretsBackend: local`?

jrajahalme replied:
I would actually favor that!

jrajahalme replied:
Done.

aanm commented on pkg/crypto/certificatemanager/certificate_manager.go:
```diff
@@ -159,3 +159,29 @@ func (m *Manager) GetTLSContext(ctx context.Context, tlsCtx *api.TLSContext) (ca
 	}
 	return "", "", "", err
 }
+
```
commit missing sign-off

jrajahalme replied:
Added.

aanm commented on pkg/crypto/certificatemanager/certificate_manager.go:
```diff
@@ -159,3 +159,29 @@ func (m *Manager) GetTLSContext(ctx context.Context, tlsCtx *api.TLSContext) (ca
 	return "", "", "", err
 }
+
+// GetSecretString returns a secret string stored in a k8s secret
```
This could have been split into a different PR for easier review.

jrajahalme replied:
Maybe review 3 first commits separately, one-by-one?

aanm commented on install/kubernetes/cilium/charts/agent/templates/clusterrole.yaml:
```diff
@@ -18,6 +18,7 @@ rules:
   - services
   - nodes
   - endpoints
+  - secrets
```
We can't blindly add this here, check my comment to make this conditional.

jrajahalme replied:
Done.

aanm commented on pkg/policy/api/l4.go:
```diff
@@ -55,6 +55,59 @@ func (p PortProtocol) Covers(other PortProtocol) bool {
+
+// Equal returns true if 'a' and 'b' have the same contents.
+func (a *Secret) Equal(b *Secret) bool {
+	return a == nil && b == nil || a != nil && b != nil && *a == *b
```
Why don't you have the receiver and the argument to be non pointers?

jrajahalme replied:
This has been deleted in a later commit in the series.

aanm commented on pkg/policy/api/l4.go:
```diff
@@ -55,6 +55,59 @@ func (p PortProtocol) Covers(other PortProtocol) bool {
+}
+
+// Equal returns true if 'a' and 'b' have the same contents.
+func (a *TLSContext) Equal(b *TLSContext) bool {
```
Same here.

jrajahalme replied:
To avoid copying the structures as would happen if they would be passed by value instead of by reference. I see that passing by value would prevent (accidental) mutations which would be nice.

jrajahalme replied:
Upon examining the call sites I recalled that the entities being compared are already pointers, so it is best to keep this `Equal()` function to avoid code repetition and keep the call sites more readable.

aanm commented on pkg/envoy/server.go:
```diff
@@ -522,34 +544,64 @@ func getHTTPRule(certManager policy.CertificateManager, h *api.PortRuleHTTP, ns
+	if len(headerMatches) == 0 {
+		headerMatches = nil
+	} else {
+		// SortHeaderMatches(headerMatches)
```
it this commented by design?

jrajahalme replied:
yes, we would optimally sort the headerMatches to avoid unnecessary updates due to different order, but that has not been implemented yet. Will also need to consider if this sorting is needed after all, as data is in a slice, so the order only changes is the order of rules in the imported policies changes? What do you think?

jrajahalme replied:
Added this comment:
```
		// Optimally we should sort the headerMatches to avoid
		// updating the policy if only the order of the rules
		// has changed. Right now, when 'headerMatches' is a
		// slice (rather than a map) the order only changes if
		// the order of the rules in the imported policies
		// changes, so there is minimal likelihood of
		// unnecessary policy updates.
```

aanm commented on pkg/envoy/server.go:
```diff
@@ -664,22 +716,27 @@ func getCiliumTLSContext(tls *policy.TLSContext) *cilium.TLSContext {
-func GetEnvoyHTTPRules(certManager policy.CertificateManager, l7Rules *api.L7Rules, ns string) *cilium.HttpNetworkPolicyRules {
+func GetEnvoyHTTPRules(certManager policy.CertificateManager, l7Rules *api.L7Rules, ns string) (*cilium.HttpNetworkPolicyRules, bool) {
 	if len(l7Rules.HTTP) > 0 { // Just cautious. This should never be false.
+		canShortCircuit := true
```
please add a comment describing what does this mean

jrajahalme replied:
Will do, how about something like this:
```
// Assume none of the rules have side-effects so that rule evaluation can
// be stopped as soon as the first allowing rule is found. 'canShortCircuit'
// is set to 'false' if any rules with side effects are encountered, causing
// all the applicable rules to be evaluated instead.
```

jrajahalme replied:
Added.

aanm commented on pkg/envoy/server.go:
```diff
@@ -709,16 +766,22 @@ func getPortNetworkPolicyRule(sel policy.CachedSelector, l7Parser policy.L7Parse
 		r.UpstreamTlsContext = getCiliumTLSContext(l7Rules.OriginatingTLS)
 	}
 
+	canShortCircuit := true
```
same here for the comment

jrajahalme replied:
will do.

aanm commented on pkg/envoy/server.go:
```diff
@@ -776,22 +839,31 @@ func getDirectionNetworkPolicy(l4Policy policy.L4PolicyMap, policyEnforced bool)
 		}
 
 		allowAll := false
+		canShortCircuit := true
```
same here for the comment

jrajahalme replied:
ditto.

aanm commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -1120,7 +1120,8 @@ var (
 		"namespace": {
 			Description: "Namespace is the namespace in which the secret exists. If " +
-				"namespace is omitted, the namespace of the enclosing rule is assumed.",
+				"namespace is omitted, the namespace of the enclosing rule is assumed, " +
```
Please bump `CustomResourceDefinitionSchemaVersion` in line 45

jrajahalme replied:
Thanks, did not know about the schema version! Will bump!

jrajahalme replied:
Done.

aanm committed:
pkg/k8s: add wrapper to get k8s secrets

Signed-off-by: André Martins <andre@cilium.io>

jrajahalme committed:
option: add 'certificates-directory' option to look for TLS certificates

This option will set the root certificate used to search for TLS
certificates defined in a CNP for L7 TLS policy enforcement.

Signed-off-by: André Martins <andre@cilium.io>

jrajahalme committed:
api: Add API support for k8s secrets

Add API definitions for k8s secrets and plumb them into Envoy policy
updates.

Add certificate manager to find certificates either locally or in k8s

Map keys "tls.key" and "tls.crt" are used for the private key and the
certificate, respectively, when creating a TLS secret like so:

$ kubectl create secret tls test-tls --key="file1" --cert="file2"

Support explicit overrides for the default item names. If given, these
must be found.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
Signed-off-by: André Martins <andre@cilium.io>

jrajahalme committed:
policy: Add PolicyContext

Add PolicyContext, a collection of references to resources needed
during policy computation. Initially this holds just a SelectorCache,
but later commits will add more items without then needing to update
and bloat all the function prototypes.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Map Envoy HTTP rules when translating policy

Supporting secrets is easier if we translate HTTP policies while in
the policy computation context. Translating HTTP policies erlier also
reduces unnecessary work of re-translating them when policy remains
but IDs change.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Translate TLS contexts into PerEpData

Resolve TLS contexts earlier in the process.

Get the default namespace from the security identity rather than the
rule. This allows for a default namespace to be well defined even if
none of the rules have a namespace label.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
test: Add runtime test for imposing an HTTP header

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
k8s: TLS testing

Add simple TLS test to access an external resource via the proxy.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Add API for actions on header mismatches

Allow replacing logging, deleting, adding, and replacing mismatching headers.

Now that we can have rules with side-effect, we must track if HTTP
rules can be short circuited. If no rules have side-effects, then the
policy evaluationn can be stopped as soon as a decision to pass
traffic has been found. If rules include side-effects, we must
evaluate all applicable rules.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

joestringer commented on pkg/crypto/certificatemanager/certificate_manager.go:
```diff
@@ -0,0 +1,196 @@
+		certPath := filepath.Join(m.rootPath, ns, name)
+		files, ioErr := ioutil.ReadDir(certPath)
+		if ioErr != nil {
+			err = fmt.Errorf("Certificates directory %s not found (%s)", certPath, ioErr)
```
nit:
```suggestion
			err = fmt.Errorf("cannot read certificates directory %s  (%s)", certPath, ioErr)
```

(If it's really `not found` then the ioErr should say this. Or permission denied, or whatever the error really is.)

jrajahalme replied:
Fixed.

joestringer commented on pkg/crypto/certificatemanager/certificate_manager.go:
```diff
@@ -0,0 +1,196 @@
+			ns = tlsCtx.Secret.Namespace
+		}
+		if tlsCtx.Secret.Name == "" {
+			err = fmt.Errorf("Missing Secret name")
```
nit: Can we provide more context in this error path, either here or in the caller? Maybe my concern is overblown, but if you have multiple TLS policies and messed up one of them, it might be a little difficult to pinpoint which one you got wrong. The caller will log directionality, but not original policy UID or namespace or anything else that might help identify which policy is miswritten.

jrajahalme replied:
Added "Required" fields to the k8s schema validation code (pkg/k8s/apis/cilium.io/v2/register.go), which takes care of this when used with k8s.

joestringer commented on pkg/crypto/certificatemanager/certificate_manager.go:
```diff
@@ -0,0 +1,196 @@
+			err = fmt.Errorf("certificates not found locally in %s nor in k8s secret %s/%s ", m.rootPath, ns, name)
+			return "", "", "", err
+		}
+		err = fmt.Errorf("certificates not found in k8s secret %s/%s", ns, name)
```
Is there any meaningful distinction between the error on line 162 and the error here, or could we unconditionally use the error on line 162 to replace lines 161-165?

It seems to me like if either local or k8s-based certificate detection was successful (and TLS is configured in the specified policy) then this function would have already returned by this point. It's the fact that neither provided certificates that leads us to this error condition.

Also the path should probably be `certPath` rather than `m.rootPath`, as that's the full path that was searched for the local certs.

jrajahalme replied:
Cleaned up in the follow-up PR #9807

joestringer commented on pkg/option/config.go:
```diff
@@ -89,6 +89,10 @@ const (
 
+	// CertsDirectory is the root directory used to find out certificates used
+	// in L7 HTTPs policy enforcement.
+	CertsDirectory = "certificates-directory"
```
nit: Should this CLI arg be more specific, like `policy-certificates-dir`, to ensure there's no mixups with the kvstore certs?

jrajahalme replied:
Maybe, will need to discuss this more. We could also generalizer this to all "secrets", not just certificates. In that case a name like "local-secrets-dir" would be a better match?

joestringer commented on pkg/envoy/server.go:
```diff
@@ -456,12 +478,34 @@ func getL7Rule(l7 *api.PortRuleL7) *cilium.L7NetworkPolicyRule {
 }
 
-func getHTTPRule(h *api.PortRuleHTTP) (headers []*envoy_api_v2_route.HeaderMatcher, ruleRef string) {
+func getHTTPRule(certManager policy.CertificateManager, h *api.PortRuleHTTP, ns string) (*cilium.HttpNetworkPolicyRule, bool) {
```
Oh, were we just ignoring the rule references string before?

:+1: if it's unused.

jrajahalme replied:
Yep, was unused.

joestringer commented on pkg/policy/repository.go:
```diff
@@ -15,22 +15,61 @@
+type PolicyContext interface {
+	GetSelectorCache() *SelectorCache
+	GetTLSContext(tls *api.TLSContext) (ca, public, private string, err error)
+	GetEnvoyHTTPRules(l7Rules *api.L7Rules) (*cilium.HttpNetworkPolicyRules, bool)
```
nit: I think it'd be useful to document particularly the `canShortcut` bool here in context of what users or implementations of this interface can and should do. The whole interface could do with a bit of documentation.

jrajahalme replied:
Added documentation to all functions in this interface.

joestringer commented on pkg/envoy/server.go:
```diff
@@ -702,22 +855,35 @@ func getDirectionNetworkPolicy(l4Policy policy.L4PolicyMap, policyEnforced bool)
 
+		// Short-circuit rules if a rule allows all and all other rules can be short-circuited
+		if allowAll && canShortCircuit {
+			log.Info("Short circuiting HTTP rules due to rule allowing all and no other rules needing attention")
```
This looks like a debug-level log, I don't think it's meaningful to most users and it's prone to repeating itself in the logs when a policy is evaluated.

jrajahalme replied:
Right, changed to `Debug`

joestringer commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -1128,10 +1169,69 @@ var (
+			Properties:  Secret,
+		},
+		"value": {
+			Description: "Value containst the header value that must be present in the request. If both Secret " +
```
```suggestion
			Description: "Value contains the header value that must be present in the request. If both Secret " +
```

jrajahalme replied:
Fixed.

joestringer commented on pkg/policy/api/utils.go:
```diff
@@ -44,6 +45,25 @@ func (h *PortRuleHTTP) Equal(o PortRuleHTTP) bool {
+		h.Value != o.Value ||
+		h.Secret != nil && o.Secret == nil ||
+		h.Secret == nil && o.Secret != nil ||
+		h.Secret != nil && o.Secret != nil && *h.Secret != *o.Secret {
```
Can this use `!h.Secret.Equal(o.Secret)`? (Ignore if this is autogenerated..)

jrajahalme replied:
`Secret.Equal()` did not exist, but I factored it out of this for clarity.

joestringer commented on pkg/policy/api/l4.go:
```diff
@@ -63,6 +102,26 @@ type PortRule struct {
+	// TerminatingTLS is the TLS context for the connection terminated by
+	// the L7 proxy.  For egress policy this specifies the server-side TLS
+	// parameters to be applied on the connections originated from the local
+	// POD and terminated by the L7 proxy. For ingress policy this specifies
```
nit: This API wording usually uses the Cilium `endpoint` terminology rather than `POD`. (Same below)

jrajahalme replied:
Changed, thanks!

joestringer reviewed (approved):
LGTM!

I have a bunch of super trivial feedback below that could be addressed in a follow-up PR, the only one that would be meaningful for the next RC would be if we wish to change the CLI arg name.
